### DISCUSSION TITLE
question about priority_threads in charon

### DISCUSSION THREAD

**Discussion by sharathbhatp** [strongswan/strongswan#2463] [Q&A] [answered]
Hi,

1. According to this document https://docs.strongswan.org/docs/5.9/config/jobPriority.html priority_threads default value is 0. Does that mean no job priority enabled? In that case can medium priority jobs (IKE_SA_INIT) use up all threads and block high priority jobs (vici command)? 

2. If above is true, if we dedicate some threads for high prio jobs using priority_threads.high, does that set a hard limit on number of threads that can be used for high prio jobs?

Thanks

**Comment by tobiasbrunner** [comment, marked as answer]
Jobs are processed based on their priority. And high priority jobs will always get processed before any medium priority jobs, which in turn get processed before any low priority jobs.

The `priority_thread` options only have an effect if there are currently no jobs of a particular priority class queued. That is, they can prevent a configured number of threads from processing lower priority jobs in order to keep them idle to immediately process jobs of a higher priority should they get queued.

And no, it's currently not a hard limit. If there are lots of high or medium priority jobs, all threads can process them and could theoretically prevent lower priority jobs from getting processed even if you configure `priority_threads.medium` or `priority_threads.low` (which, as documented, has currently no effect). So yeah, these config options have some limitations.
